appinventor/components/src/com/google/appinventor/components/runtime/Navigation.java

```diff
@@ -0,0 +1,352 @@
+import com.google.appinventor.components.annotations.SimpleProperty;
+import com.google.appinventor.components.common.ComponentCategory;
+import com.google.appinventor.components.common.PropertyTypeConstants;
+import com.google.appinventor.components.common.YaVersion;
+import com.google.appinventor.components.runtime.util.AsynchUtil;
+import com.google.appinventor.components.runtime.util.ErrorMessages;
+import com.google.appinventor.components.runtime.util.GeoJSONUtil;
+import com.google.appinventor.components.runtime.util.JsonUtil;
+import com.google.appinventor.components.runtime.util.MapFactory.MapFeatureContainer;
+import com.google.appinventor.components.runtime.util.MapFactory.MapLineString;
+import com.google.appinventor.components.runtime.Web;
+
+import static com.google.appinventor.components.runtime.util.GeometryUtil.isValidLatitude;
+import static com.google.appinventor.components.runtime.util.GeometryUtil.isValidLongitude;
+
+import android.util.Log;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.BufferedOutputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.ArrayList;
+
+@DesignerComponent(version = YaVersion.NAVIGATION_COMPONENT_VERSION,
+    category = ComponentCategory.MAPS,
+    description = "Navigation",
+    nonVisible = true)
+@UsesPermissions(permissionNames = "android.permission.INTERNET")
+@UsesLibraries(libraries = "json.jar")
+@SimpleObject
+public class Navigation extends AndroidNonvisibleComponent implements Component {
+
+  private static final String TAG = "Navigation";
+
+  public static final String OPEN_ROUTE_SERVICE_URL = "https://api.openrouteservice.org/v2/directions/%s/geojson/";
+  private final Activity activity;
+  private String apiKey;
+  private GeoPoint startLocation;
+  private GeoPoint endLocation;
+  private TransportMethod method;
+  private Web web;
+
+  enum TransportMethod {
+    DEFAULT ("foot-walking"),
+    DRIVING ("driving-car"),
+    CYCLING ("cycling-regular"),
+    WALKING ("foot-walking"),
+    WHEELCHAIR ("wheelchair");
+
+    private final String method;
+    
+    TransportMethod(String method) {
+      this.method = method;
+    }
+    
+    private String method() { 
+      return method;
+    }
+  }
+
+  /**
+   * Creates a Navigation component.
+   *
+   * @param container container, component will be placed in
+   */
+  public Navigation(ComponentContainer container) {
+    super(container.$form());
+    activity = container.$context();
+    apiKey = "";
+    startLocation = new GeoPoint(0, 0);
+    endLocation = new GeoPoint(0, 0);
+    method = TransportMethod.DEFAULT;
+  }
+
+  @SimpleFunction(description = "")
+  public void RequestDirections() {
+    AsynchUtil.runAsynchronously(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          performRequest();
```

Comment:
It's possible that the developer might make multiple requests in the same block stack. Therefore, there's no guarantee that by the time this code executes on a different thread that they haven't changed the start, end, or method of direction. Instead, copy the method, start and end coordinates (make them `final` in this function prior to the runAsynchronously call), and use them in the body of `performRequest` rather than referencing the ones declared on `this`.